### PR TITLE
Fixing our DoublyLinkedList

### PR DESCRIPTION
**Pull request by castilma**
Our (currently unused) DoublyLinkedList had some real issues.
While writing tests, I discovered many things, but the real interesting things were those, that should have raised a compilation error, but didn't. I guess they didn't, because we defined everything in a header and the functions containing uncompilable code were never called(referenced).

I'm not exactly sure, what we can learn from this.
I would suggest using header files for declaration and a separate file for definition. When trying to create `doubly_linked_list.o` from `doubly_linked_list.cpp`, that should have raised an error.
Besides that: If we keep everything in one header, doesn't that lead to every compilation unit recreating the machine code for that class?

### CODE REVIEW

File: libopenage/datastructure/doubly_linked_list.h [hunk 1]
```diff
@@ -85,6 +85,10 @@ class DoublyLinkedList {
 
 		node->next->previous = new_node;
 		node->next           = new_node;
+
+		if (this->last == node) {
+				this->last = new_node;
+		}
```
Comment (castilma) [inline libopenage/datastructure/doubly_linked_list.h]: like insert_after(), I'm not sure, if we want to make insert_*() check for this, or let the caller take responsibility for that case.

File: libopenage/datastructure/doubly_linked_list.h [hunk 2]
```diff
@@ -121,14 +129,14 @@ class DoublyLinkedList {
...
+		if (this->size() > 0) {
 			do {
 				if (current->data == item) {
-					this->erase_node(current);
+					this->erase(current);
 					return;
 				}
 				current = current->next;
-			} while (current != this->end);
+			} while (current != this->last);
```
Comment (castilma) [inline libopenage/datastructure/doubly_linked_list.h]: `last` and `first` maybe should be callled `back` and `front`.

File: libopenage/datastructure/doubly_linked_list.h [hunk 3]
```diff
@@ -240,14 +248,17 @@ class DoublyLinkedList {
 	/**
 	 * Delete all list nodes.
 	 * O(n)
+	 * asdf: don't call erase: this safes relinking all nodes.
 	 */
 	void clear() {
 		node_t *delete_now = this->first;
 		while (this->node_count > 0) {
-			node_t *deleted_next = this->first->next;
```
Comment (castilma) [inline libopenage/datastructure/doubly_linked_list.h]: this is just wrong.

File: libopenage/datastructure/doubly_linked_list.h [hunk 4]
```diff
@@ -114,28 +118,42 @@ class DoublyLinkedList {
...
 	/**
 	 * Fully delete a node and fix the neighbor pointers.
+	 * Update #first or #last, if necessary.
 	 * O(1)
+	 * @pre \p node belongs to this list.
 	 */
 	void erase(node_t *node) {
-		node->previous->next = node->next;
-		node->next->previous = node->previous;
+		assert(this->node_count != 0);
```
Comment (castilma) [inline libopenage/datastructure/doubly_linked_list.h]: remove this assert?

File: libopenage/datastructure/doubly_linked_list.h [hunk 5]
```diff
@@ -114,28 +118,42 @@ class DoublyLinkedList {
 
 	/**
 	 * Fully delete a node and fix the neighbor pointers.
+	 * Update #first or #last, if necessary.
 	 * O(1)
+	 * @pre \p node belongs to this list.
 	 */
 	void erase(node_t *node) {
```
Comment (castilma) [inline libopenage/datastructure/doubly_linked_list.h]: Do we want the erase() functions to correct the first/last pointer if necessary? 
there can be a protected unlink() function, which does the same without setting first and last.

**Comment by TheJJ** [comment]
I added that list back when I was writing the heap, mainly as i was toying with fibonacci heaps. If it is unused now, we should just remove it?

**Comment by castilma** [comment]
I'm a fan of removing code. I'd say yes.

**Comment by TheJJ** [comment]
K, go for it :)

**Comment by TheJJ** [comment]
Fixed by removal :stuck_out_tongue:

**Comment by zuntrax** [comment]
Fix by removal is best fix indeed.

**Pull request merged**